src/time/Waveform.cpp

```diff
@@ -31,8 +32,17 @@ void Waveform::store(const Eigen::VectorXd &data)
   this->_timeWindows.col(columnID) = data;
 }
 
+Eigen::VectorXd Waveform::sample(double dt, int timeWindows, int order)
+{
+  PRECICE_ASSERT(dt >= 0, "Sampling outside of valid range!");
+  PRECICE_ASSERT(dt <= 1, "Sampling outside of valid range!");
```

Comment:
There might actually be a bug in the current implementation. We have to be careful here: What happens, if the time-window size is different from 1? So far our tests use a time window size of 1 and we should also test the more general case. Otherwise it is easy to miss errors.

[This code](https://github.com/BenjaminRodenberg/precice/pull/1/files#diff-cf9df0e4694b35c558b21d42de07db93a0bf82ec41a311e54350a5fafe1306d6R1238) shows the intended use of the function:

1. User provides `dt` (timeStepDt) relative to start of the current time step via `read...Data` function. (User does not want to worry about the window, but only about the current step. So `0` is always pointing to the beginning of the time step)
2. Inside of the `read...Data` function the `dt` given by the user is mapped to a `timeWindowDt`, where `0` refers to the beginning of the window. (preCICE knows about the window and our waveform also lives on the window)
3. Data context is used to sample the waveform at the corresponding point in time. `DataContext::sampleAt` just calls `_providedWaveform->sample()`.

In step 3 there might be a bug: `timeWindowDt` is restricted to `[0, timeWindowSize]`, but `sample` expects a `dt`, which is normalized to `[0,1]` (`normalizedDt`). I'm a bit unsure here. On the one hand it looks like an overkill to use three different `dt`s (`timeStepDt`, `timeWindowDt` and `normalizedDt`), but on the other hand every dt seems to be the "right" dt for each context:

* A user can only know the `timeStepDt`, if the position inside the time window is not tracked. 
* preCICE knows where the current time step is located within the window and has to map the local `timeStepDt` to the global `timeWindowDt`. 
* The Waveform does (at least for now) not have to worry about the size of the window. This simplifies the class. Mapping to the fixed interval `[1,0]` simplifies the implementation and allows to check input arguments easily.

I will now start with using the more explicit names `timeStepDt`, `timeWindowDt` and `normalizedDt` in this PR.

Reply:
I changed the names in b50d1341a9b6d60c078ed2ba1bb3673ec8bd302c.

There was indeed a bug in my other PR (https://github.com/BenjaminRodenberg/precice/pull/1). See https://github.com/BenjaminRodenberg/precice/pull/1/commits/6db926f073731f081b08a3b0aa9ac8a046556a02 for what can go wrong.